src/haskell/duck.hs

```diff
@@ -0,0 +1,41 @@
+{- A typeclass that Wolf and Mallard will have to explicitly instantiate -}
+class Duck a where
+    quake :: a -> String
+    dance :: a -> String
+
+data Wolf = Wolf
+
+eat :: (Duck t) => Wolf -> t -> String
+eat _ _ = " 😈 "
+
+{- In order to become a duck, Wolf has to use the Adapter pattern. In
```

Comment:
Indeed, because Haskell does not apply coercion by default. For example with ocaml, I needed to explicitly call the coercion operator to check and convert a wolf as a duck (thewolf :> duck). Maybe something like that exists in Haskell but I cannot find how to do that.
Can you check whether https://wiki.haskell.org/GHC/Coercible can solve the problem?

Cheers
(but do not forget to work your exams...)